### PR TITLE
githooks: init at 3.0.1

### PR DESCRIPTION
**Pull request by gabyx**
## Description of changes

This adds [Githooks](https://github.com/gabyx/Githooks) at version 3.0.1 which is a platform-independent Git hooks manager.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gabyx** [comment]
I am not sure what package tests I should do. I wont do any Githooks interal tests, as those are already done upstream in the repo, is that ok?

**Review by Aleksanaa — COMMENTED** [review]
Recommend to format with `nixfmt-rfc-style`, not enforced now but no spaces at both ends of the package name will affect our ability to use grep like tools 🫠

**Comment by Aleksanaa** [comment]
Also please reword commit to `githooks: init at 3.0.1`

**Comment by Aleksanaa** [comment]
> I am not sure what package tests I should do. I wont do any Githooks interal tests, as those are already done upstream in the repo, is that ok?

That's okay but why did you disable them?

File: pkgs/by-name/gi/githooks/package.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  modRoot = "./githooks";
+  vendorHash = "sha256-ZcDD4Z/thtyCvXg6GzzKC/FSbh700QEaqXU8FaZaZc4=";
+  nativeBuildInputs = [makeWrapper];
+  buildInputs = [git];
+
+  ldflags = [
+    "-s" # Disable symbole table.
+    "-w" # Disable DWARF generation.
+  ];
```
Comment (AndersonTorres) [inline pkgs/by-name/gi/githooks/package.nix:29]:
```suggestion
  ldflags = [
    "-s" "-w"
  ];
```

File: pkgs/by-name/gi/githooks/package.nix [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    package = "githooks-cli";
+    command = "githooks-cli --version";
+    inherit version;
+  };
+
+  meta = with lib; {
+    description = "Githooks is a Git hooks manager with per-repo and shared Git hooks including version control";
+    homepage = "https://github.com/gabyx/Githooks";
+    license = licenses.mpl20;
+    maintainers = with maintainers; [gabyx];
```
Comment (AndersonTorres) [inline pkgs/by-name/gi/githooks/package.nix:60]:
```suggestion
    maintainers = with maintainers; [ gabyx ];
```

File: pkgs/by-name/gi/githooks/package.nix [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  lib,
+  buildGo122Module,
+  fetchFromGitHub,
+  git,
+  testers,
+  makeWrapper,
+}:
+buildGo122Module rec {
+  pname = "githooks";
+
```
Comment (AndersonTorres) [inline pkgs/by-name/gi/githooks/package.nix:11]:
```suggestion
```

File: pkgs/by-name/gi/githooks/package.nix [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    owner = "gabyx";
+    repo = "githooks";
+    rev = "v${version}";
+    hash = "sha256-qv0nl3EEYVo/s79r+yK3ZQCGPXM2bzGdWatPY24aOZg=";
+  };
+
+  modRoot = "./githooks";
+  vendorHash = "sha256-ZcDD4Z/thtyCvXg6GzzKC/FSbh700QEaqXU8FaZaZc4=";
+  nativeBuildInputs = [makeWrapper];
+  buildInputs = [git];
```
Comment (AndersonTorres) [inline pkgs/by-name/gi/githooks/package.nix:24]:
```suggestion

  nativeBuildInputs = [ makeWrapper ];

  buildInputs = [ git ];

  strictDeps = true;
```

File: pkgs/by-name/gi/githooks/package.nix [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  ldflags = [
+    "-s" # Disable symbole table.
+    "-w" # Disable DWARF generation.
+  ];
+
+  tags = ["package_manager_enabled"];
+
+  doCheck = false;
+
+  postConfigure = ''
```
Comment (AndersonTorres) [inline pkgs/by-name/gi/githooks/package.nix:35]: Is this correct?

File: pkgs/by-name/gi/githooks/package.nix (hunk 1 above)
Comment (gabyx) [reply pkgs/by-name/gi/githooks/package.nix:29]: I like these comments, because its not obvious.

File: pkgs/by-name/gi/githooks/package.nix (hunk 5 above)
Comment (gabyx) [reply pkgs/by-name/gi/githooks/package.nix:35]: Yes we need to configure some build files before building.

File: pkgs/by-name/gi/githooks/package.nix [hunk 6]
```diff
@@ -0,0 +1,63 @@
+{
+  lib,
+  buildGo122Module,
+  fetchFromGitHub,
+  git,
+  testers,
+  makeWrapper,
+}:
+buildGo122Module rec {
```
[the remark this answers is not in the document]
Comment (gabyx) [reply pkgs/by-name/gi/githooks/package.nix:9]: ok, good to know: but 1.22 is the hard requirement for Githooks, with that it was tested.

**Comment by gabyx** [comment]
@Aleksanaa : Thanks for the review, applied all changes, formatted also and repushed.

File: pkgs/by-name/gi/githooks/package.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (Aleksanaa) [reply pkgs/by-name/gi/githooks/package.nix:9]: Nixpkgs only maintains supported versions of golang so pining a version unconditionally will only increase workload for upgrades. Golang is assumed to be backward compatible anyway.

File: pkgs/by-name/gi/githooks/package.nix [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  buildInputs = [git];
+
+  ldflags = [
+    "-s" # Disable symbole table.
+    "-w" # Disable DWARF generation.
+  ];
+
+  tags = ["package_manager_enabled"];
+
+  doCheck = false;
```
[the remark this answers is not in the document]
Comment (Aleksanaa) [reply pkgs/by-name/gi/githooks/package.nix:41]: Have you run the tests? Do them work on your machine? Can you only skip tests that need to be containerized?

File: pkgs/by-name/gi/githooks/package.nix [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      GH_BUILD_TAG="v${version}" \
+      go generate -mod=vendor ./...
+  '';
+
+  postInstall = ''
+    mv "$out/bin/cli" "$out/bin/githooks-cli"
+    mv "$out/bin/runner" "$out/bin/githooks-runner"
+    mv "$out/bin/dialog" "$out/bin/githooks-dialog"
+
+    wrapProgram "$out/bin/githooks-cli" --prefix PATH : ${lib.makeBinPath [git]}
```
[the remark this answers is not in the document]
Comment (Aleksanaa) [reply pkgs/by-name/gi/githooks/package.nix]: Sorry, I mean the wrappers should be generated in `postFixup`, but the `mv` should be in `postInstall`. Also using `install` will be better than `mv`.

[the remark this answers is not in the document]
Comment (gabyx) [reply pkgs/by-name/gi/githooks/package.nix]: Thanks!

File: pkgs/by-name/gi/githooks/package.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (gabyx) [reply pkgs/by-name/gi/githooks/package.nix]: @Aleksanaa Is doCheck for testing the package or testing Githooks?
If it needs tests, than jeah it needs more dependencies, `git-lfs` is optional etc. I have only containerized integrations tests and 
Go unit tests (but only few which is not of much use I guess), I added running the tests in my last commit.

Is it possible to depend on `docker/podman` but only for the tests? I tried adding `podman` to the function input and `nativeCheckInputs = [podman]`, which did not work for the tests. However so far I guess this should be fine.

Also I discovered some wrong build stuff in `git-town` (wrong regex) which I need to report.

File: pkgs/by-name/gi/githooks/package.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (gabyx) [reply pkgs/by-name/gi/githooks/package.nix:9]: Ok, thanks, so I use the normal one.

**Comment by gabyx** [comment]
@Aleksanaa : Thanks for the review, I made the changes . Added the Go tests too.

File: pkgs/by-name/gi/githooks/package.nix [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  preCheck =
+    let
+      skippedTests = [
+        "TestGithooksCompliesWithGit" # Needs internet to download all hooks documentation.
+        "TestUpdateImages" # Needs docker/podman.
+        ];
+    in
+    ''
+      # Disable some tests.
+      checkFlags="-v -skip (${builtins.concatStringsSep "|" skippedTests})"
```
Comment (Aleksanaa) [inline pkgs/by-name/gi/githooks/package.nix:47]: Not needed in preCheck, you can write it directly as a derivation attribute

File: pkgs/by-name/gi/githooks/package.nix [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+  # We need to generate some build files before building.
+  postConfigure = ''
+    GH_BUILD_VERSION="${version}" \
+      GH_BUILD_TAG="v${version}" \
+      go generate -mod=vendor ./...
+  '';
+
+  postInstall = ''
+    install -Dm755 $out/bin/cli $out/bin/githooks-cli
```
Comment (Aleksanaa) [inline pkgs/by-name/gi/githooks/package.nix:60]: This is equal to `cp`. I guess we don't want to have two executables for the same command, so we need `mv`?

File: pkgs/by-name/gi/githooks/package.nix (hunk 9 above)
Comment (Aleksanaa) [reply pkgs/by-name/gi/githooks/package.nix:47]: It's a list. Please search `checkFlags` in nixpkgs manual.

File: pkgs/by-name/gi/githooks/package.nix (hunk 10 above)
Comment (gabyx) [reply pkgs/by-name/gi/githooks/package.nix:60]: Ou, ups. right! thanks, stupid mistake.

**Comment by gabyx** [comment]
@Aleksanaa: Thanks for the last review hopefully. I oriented myself on git-town, which might get some improvements after we merged this. Git-Town nix file has some unpleasantnes  I guess in the build code too...

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**